docs/design/database.md

```diff
@@ -0,0 +1,133 @@
+1. Once the device is registered a `POST /zones/$id/peers` was issued to create an association between this device and a zone.
+1. `GET /zones/$id/peers` was polled to retrieve the peer listing for a zone.
+
+On top of this flow, we were somewhat abusing `POST /zones/$id/peers` to also update peers after NAT-T.
+
+This model was built under the assumption that multi-tenancy would be achieved by deploying an Apex stack per-tenant.
+
+# v2: SaaS (Milestone-1)
+
+In order to create a more SaaS-ready schema, we've made the following changes.
+
+1. Replace the `Zone` construct with `Organization`. This is a GitHub like construct, where each unique `User` is also it's own `Organization`.
+1. Deprecate `Peer` and combine it's functionality with the `Device` table.
+
+```mermaid
+erDiagram
+    USER ||--o{ DEVICE : has
+    USER ||--|{ ORGANIZATION : belongs_to
+    ORGANIZATION ||--o{ DEVICE: contains
+    USER{
+        string id
+        string username
+    }
+    DEVICE{
+        string id
+        string user_id
+        string organization_id
+        string public_key
+        string hostname
+        string tunnel_ip
+        string allowed_ips
+        string local_ip
+        string child_prefix
+        string relay
+        string organization_prefix
+        string reflexive_ipv4
+        string endpoint_local_address_ipv4
+        string symmetric_nat
+    }
+    ORGANIZATION{
+        string id
+        string name
+        string description
+        string cidr
+        string hub_zone
+    }
+```
+
+In this simplified model we only have 3 concepts:
+
+- Users: Unique users of the service
+- Devices: A unique end-user device
+- Organization: A collection of users who have permissions to onboard one or more devices into this organization
+
+Since a `User` is also an `Organization`, once someone has registered to Apex we onboard devices into their personal organization.
+
+The device-onboarding flow is as follows:
+
+1. A User registers for Apex using an sign-up form on https://apex-hosted.cloud, with the option to use SSO from Google, Github, Facebook etc..
+1. User downloads and runs the Apex agent
+1. Apex agent does a `POST /device` to register the unique device with Apex for the user - it's onboarded into their personal organization.
+1. `GET /organization/$id/devices` is polled to retrieve the peer listing for a zone.
+1. An dedicated API endpoint was added to update device information if the public-facing address changes etc...
+
+A key point here is that there is no admin intervention required to onboard a new user as anyone can sign-up and register.
+
+Supporting a user belonging to multiple organizations is easy in this model since a `Device` may only be in one `Organization` at a time. A `User` may chose to move their device at any time.
```

Comment:
... but actually supporting this isn't expected at first, right?  You could call the "multi-organization" support v2.5 or something?

Reply:
I've added a note about this in the text that the functionality won't be implemented in v1->v2 but that the schema supports such a feature. It wouldn't require changes so we wouldn't be at v2.5.